crates/astria-sequencer-relayer/src/relayer/celestia_client/mod.rs

```diff
@@ -0,0 +1,654 @@
+pub(super) struct CelestiaClient {
+    /// The inner `tonic` gRPC channel shared by the various generated gRPC clients.
+    grpc_channel: Channel,
+    /// A gRPC client to broadcast and get transactions.
+    tx_client: TxClient<Channel>,
+    /// A gRPC client to query for blob params (we're interested in the `gas_per_blob_byte`).
+    blob_query_client: BlobQueryClient<Channel>,
+    /// A gRPC client to query for general account params (we're interested in the
+    /// `tx_size_cost_per_byte`) and for values specific to our account, namely the account number
+    /// and sequence number (i.e. nonce).
+    auth_query_client: AuthQueryClient<Channel>,
+    /// The crypto keys associated with our Celestia account.
+    signing_keys: CelestiaKeys,
+    /// The Bech32-encoded address of our Celestia account.
+    address: Bech32Address,
+    /// The Celestia network ID.
+    chain_id: String,
+}
+
+impl CelestiaClient {
+    /// Tries to submit the given blobs to the Celestia app.
+    ///
+    /// The `last_error_receiver` will provide the error from the previous attempt if this is not
+    /// the first attempt for these blobs, or `None` if it is the first attempt.  The error can be
+    /// used to obtain the appropriate fee in the case that the previous attempt failed due to a
+    /// low fee.
+    // Copied from https://github.com/celestiaorg/celestia-app/blob/v1.4.0/x/blob/payforblob.go
+    pub(super) async fn try_submit(
+        mut self,
+        blobs: Arc<Vec<Blob>>,
+        state: Arc<super::State>,
+        last_error_receiver: watch::Receiver<Option<Error>>,
+    ) -> Result<u64, Error> {
+        // Get the error from the last attempt to `try_submit` and if `Some`, just update the
+        // cached cost params.
+        let maybe_last_error = last_error_receiver.borrow().clone();
+        if maybe_last_error.is_some() {
+            self.fetch_and_cache_cost_params(state.clone()).await?;
+        }
+
+        let msg_pay_for_blobs = new_msg_pay_for_blobs(blobs.iter(), self.address.clone())?;
+
+        let base_account = self.fetch_account().await?;
+        debug!(
+            account_number = base_account.account_number,
+            sequence = base_account.sequence,
+            "got account info"
+        );
+
+        let cost_params = state.celestia_cost_params();
+        let gas_limit = estimate_gas(&msg_pay_for_blobs.blob_sizes, cost_params);
+        let fee = calculate_fee(cost_params, gas_limit, maybe_last_error);
+        trace!(?cost_params, gas_limit = gas_limit.0, fee_utia = fee);
+        debug!(fee_utia = fee, "estimated fee");
+
+        let signed_tx = new_signed_tx(
+            &msg_pay_for_blobs,
+            &base_account,
+            gas_limit,
+            fee,
+            self.chain_id.clone(),
+            &self.signing_keys,
+        );
+
+        let blob_tx = new_blob_tx(&signed_tx, blobs.iter());
+
+        let tx_hash = self.broadcast_tx(blob_tx).await?;
+
+        loop {
```

Comment:
maybe replace with a `broadcast_tx_commit` call with a timeout, as it looks like this is waiting for the tx to be included in a block?

Reply:
I believe there may be issues around that, since [the ticket](https://github.com/astriaorg/astria/issues/921) specified using sync mode for this.

Reply:
We don't want to use broadcast_tx_commit because it doesn't actually wait for the data to be included, it waits up to 10 seconds which is less than a celestia block time the cometbft rpc docs note: "IMPORTANT: use only for testing and development. In production, use BroadcastTxSync or BroadcastTxAsync. You can subscribe for the transaction result using JSONRPC via a websocket. See https://docs.cometbft.com/v0.38.x/core/subscription.html"

Reply:
@Fraser999 Can you add this bit that @joroshiba linked above that line? This strikes me as the kind of thing that will come up in 6 months from now where we will be wondering anew why we decided to use `broadcast_tx` instead of `broadcast_tx_commit`. Note that @joroshiba's link appears to be broken 11h later. This one still works: https://docs.cometbft.com/v0.38/rpc/#/Tx/broadcast_tx_commit

And here are full permalinks that say the same thing:
https://github.com/cometbft/cometbft/blob/b139e139ad9ae6fccb9682aa5c2de4aa952fd055/rpc/openapi/openapi.yaml#L201
https://github.com/cometbft/cometbft/blob/b139e139ad9ae6fccb9682aa5c2de4aa952fd055/docs/references/config/config.toml.md?plain=1#L571

Reply:
Done in [97d0f96](https://github.com/astriaorg/astria/pull/963/commits/97d0f960fd5f54c2312097777f2d5cdd3400f3d4).